QModManager/API/SMLHelper/Assets/ModPrefab.cs

```diff
@@ -0,0 +1,123 @@
+
+        /// <summary>
+        /// The <see cref="TechType"/> of the corresponding item.
+        /// Used for <see cref="Fixer" />, <see cref="TechTag" />, and <see cref="Constructable" /> components whenever applicable.
+        /// </summary>
+        public TechType TechType { get; protected set; }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ModPrefab" /> class.
+        /// </summary>
+        /// <param name="classId">The class identifier used for the <see cref="PrefabIdentifier" /> component whenever applicable.</param>
+        /// <param name="prefabFileName">Name of the prefab file.</param>
+        /// <param name="techType">The techtype of the corresponding item. 
+        /// Used for the <see cref="Fixer" />, <see cref="TechTag" />, and <see cref="Constructable" /> components whenever applicable.
+        /// Can also be set later in the constructor if it is not yet provided.</param>
+        protected ModPrefab(string classId, string prefabFileName, TechType techType = TechType.None)
+        {
+            ClassID = classId;
+            PrefabFileName = prefabFileName;
+            TechType = techType;
+
+            ModName = this.GetType().Assembly.GetName().Name;
+        }
+
+        internal GameObject GetGameObjectInternal()
+        {
+            if (cachedPrefab)
+            {
+                // This will avoid GetGameObject() being called every time the object is spawned. Could break things if there is a degree 
+                // of random generation in GetGameObject(), in that case the modder will need to do their random stuff in a MonoBehaviour.Awake()
+                // e.g. in MinecraftFish, the tropical fish have random colours which is done in a MonoBehaviour, rather than the GetGameObject() method
+                return cachedPrefab;
+            }
+
+            GameObject go = GetGameObject();
+
+            if (go == null)
+            {
+                return null;
+            }
+
+            go.transform.position = new Vector3(-5000, -5000, -5000);
+            go.name = ClassID;
+
+            /* Make sure prefab doesn't get cleared when quiting game to menu. */
+            SceneCleanerPreserve scp = go.AddComponent<SceneCleanerPreserve>();
+            scp.enabled = true;
+          
+            if (TechType != TechType.None)
+            {
+                go.AddComponent<Fixer>().techType = TechType;
+
+                if (go.GetComponent<TechTag>() != null)
+                {
+                    go.GetComponent<TechTag>().type = TechType;
+                }
+
+                if (go.GetComponent<Constructable>() != null)
+                {
+                    go.GetComponent<Constructable>().techType = TechType;
+                }
+            }
+
+            if (go.GetComponent<PrefabIdentifier>() != null)
+            {
+                go.GetComponent<PrefabIdentifier>().ClassId = ClassID;
```

Comment:
I have come across instances where a modded prefab I created did not have the `PrefabIdentifier` component when cloned, making it so that the component had to be added manually in the `override GetGameObject` method.  
Without this component, there isn't really a good way to identify the item when save data is being loaded.

Open question:  
**Should we go ahead and ensure that this component is always added to all `ModPrefab` created gameobjects?**

Reply:
I haven't worked a lot with prefabs, so I'll let you handle this one